spotify/models/user.py

```diff
@@ -207,77 +291,112 @@ def _raise(*args, **kwargs):
+            token_info: TokenInfo,
+            cache_path: Optional[Union[Path, str]] = None,
+            refresh: bool = False
+    ) -> "spotify.User":
+        """Create a :class:`User` object from token information.
+        Either from a provided access or refresh token.
 
-            coroutine: Coroutine = self._refreshing_token(
-                expires_in, refresh_token
-            )  # pylint: disable=protected-access
+        Parameters
+        ----------
+        client : :class:`spotify.Client`
+            The spotify client to associate the user with.
+        token_info : :class:`TokenInfo`
+            The token information to authenticate the User object
+            with. Contains an access_token with corresponding
+            expires_at information in seconds, refresh_token
+            or both.
+        cache_path: Optional[Union[:class:`Path`, :class:`str`]]
+            Optional path to a cache file where tokens and refresh
+            tokens will be saved.
+        refresh : Optional[Tuple[:class:`int`, :class:`str`]]
+            When provided the refresh argument must be a tuple
+            of an integer representing the number of seconds until
+            the access token expires and a string representing the
+            refresh token to use to generate a new access token.
+        """
+        if cache_path and isinstance(cache_path, str):
+            cache_path = Path(cache_path)
 
-            self._refresh_task = self.client.loop.create_task(
-                coroutine
-            )  # pylint: disable=protected-access
+        if not token_info.is_token_valid():
+            # expired access token, need to use refresh token
+            if not token_info.refresh_token:
+                raise Exception("Invalid access token and no refresh token provided")
+
+            headers = {
+                "Authorization":
+                    f"Basic {b64encode(':'.join((client.http.client_id, client.http.client_secret)).encode()).decode()}",
+                "Content-Type": "application/x-www-form-urlencoded",
+            }
+            route = ("POST", REFRESH_TOKEN_URL.format(refresh_token=token_info.refresh_token))
+            token_info = TokenInfo(**await client.http.request(route, headers=headers))
 
-        return self
+        http = HTTPUserClient(token_info.access_token)
+
+        data = await http.current_user()
+
+        return cls(
+            client,
+            data=data,
+            http=http,
+            token_info=token_info,
+            refresh=refresh,
+            cache_path=cache_path
+        )
 
     @classmethod
-    async def from_refresh_token(
-        cls, client: "spotify.Client", refresh_token: str, refresh: bool = False, *args
-    ):
-        """Create a :class:`User` object from a refresh token.
-        It will poll the spotify API for a new access token and
-        use that to initialize the spotify user.
+    async def from_cache(
```

Comment:
depending on if `TokenInfo` is intended to be a public model, this constructor may not have to exist.